crates/biome_js_analyze/src/lint/nursery/no_substr.rs

```diff
@@ -0,0 +1,204 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, ActionCategory, Ast, FixKind, Rule, RuleDiagnostic,
+};
+use biome_console::markup;
+use biome_js_factory::make;
+use biome_js_syntax::{
+    AnyJsName, JsCallArguments, JsExpressionStatement, JsSyntaxToken, JsVariableStatement,
+};
+use biome_rowan::{declare_node_union, AstSeparatedList, BatchMutationExt, TextRange, TokenText};
+
+use crate::JsRuleAction;
+
+declare_rule! {
+    /// Enforce the use of `String.slice()` over `String.substr()` and `String.substring()`.
+    ///
+    /// `String.slice()` is preferred over `String.substr()` and `String.substring()` because it is a more popular option with clearer behavior,
+    ///  and it has a consistent counterpart in arrays.
+    ///
+    /// Note that `String.substr`, `String.substring` and `String.slice` are not identical when arguments are passed.
+    /// For detailed differences, refer to the MDN documentation:
+    /// - [The difference between substring() and substr()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#the_difference_between_substring_and_substr)
+    /// - [Differences between substring() and slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#differences_between_substring_and_slice)
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substr();
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substring();
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// foo.slice(beginIndex, endIndex);
+    /// ```
+    ///
+    pub NoSubstr {
+        version: "next",
+        name: "noSubstr",
+        language: "js",
+        recommended: false,
```

Comment:
You forgot to add the source.

```suggestion
        recommended: false,
        sources: &[RuleSOurce::EslintUnicorn("prefer-string-slice")],
```

Note: don't forget to run `just gen-lint`.

Reply:
Ok, I will add the source.